File: SOURCE/EFEX/BASE/JAVA/cMailbox.java
/**
 * Package : ISI Efex
 * Type    : Class
 * Name    : cMailbox
 * Author  : Steve Gregan
 * Date    : June 2008
 */
package com.isi.efex;
import java.util.Vector;

/**
 * This class defines a mailbox that can be used for messaging
 */
public final class cMailbox {

   //
   // Class static variables
   //
   private static final String BUFFER_OPEN = "<EFEX>";
   private static final String BUFFER_CLOSE = "</EFEX>";
   
   //
   // Request constants
   //
   public static final char EFEX_RQS_USERNAME = 101;
   public static final char EFEX_RQS_PASSWORD = 102;
   public static final char EFEX_RQS_CUSTOMER_ID = 103;
   public static final char EFEX_RQS_MESSAGE = 199;

   //
   // Mobile constants
   //
   public static final char EFEX_MOB = 501;
   public static final char EFEX_MOB_USERNAME = 502;
   public static final char EFEX_MOB_LANGUAGE = 503;
   public static final char EFEX_MOB_SERVERURL = 504;
   public static final char EFEX_MOB_SECURE = 505;
   
   //
   // Control constants
   //
   public static final char EFEX_CTL_STR = 1000;
   //
   public static final char EFEX_CTL = 1001;
   public static final char EFEX_CTL_USER_FIRSTNAME = 1002;
   public static final char EFEX_CTL_USER_LASTNAME = 1003;
   public static final char EFEX_CTL_MOBILE_DATE = 1004;
   public static final char EFEX_CTL_MOBILE_STATUS = 1005;
   public static final char EFEX_CTL_MOBILE_LOADED_TIME = 1006;
   public static final char EFEX_CTL_MOBILE_SAVED_TIME = 1007;
   //
   public static final char EFEX_CTL_END = (char)(1999);

   //
   // Route constants
   //
   public static final char EFEX_RTE_STR = 2000;
   //
   public static final char EFEX_RTE_CALL = 2001;
   public static final char EFEX_RTE_CALL_SEQUENCE = 2002;
   public static final char EFEX_RTE_CALL_CUSTOMER_ID = 2003;
   public static final char EFEX_RTE_CALL_CUSTOMER_CODE = 2013;
   public static final char EFEX_RTE_CALL_CUSTOMER_NAME = 2004;
   public static final char EFEX_RTE_CALL_CUSTOMER_TYPE = 2005;
   public static final char EFEX_RTE_CALL_MARKET = 2006;
   public static final char EFEX_RTE_CALL_STATUS = 2007;
   public static final char EFEX_RTE_CALL_DATE = 2008;
   public static final char EFEX_RTE_CALL_STR_TIME = 2009;
   public static final char EFEX_RTE_CALL_END_TIME = 2010;
   public static final char EFEX_RTE_CALL_ORDER_SEND = 2011;
   public static final char EFEX_RTE_CALL_STOCK_DIST_COUNT = 2012;
   //
   public static final char EFEX_RTE_STCK_ITEM = 2101;
   public static final char EFEX_RTE_STCK_ITEM_ID = 2102;
   public static final char EFEX_RTE_STCK_ITEM_NAME = 2103;
   public static final char EFEX_RTE_STCK_ITEM_REQUIRED = 2104;
   public static final char EFEX_RTE_STCK_ITEM_QTY = 2105;
   //
   public static final char EFEX_RTE_ORDR_ITEM = 2201;
   public static final char EFEX_RTE_ORDR_ITEM_ID = 2202;
   public static final char EFEX_RTE_ORDR_ITEM_NAME = 2203;
   public static final char EFEX_RTE_ORDR_ITEM_PRICE_TDU = 2204;
   public static final char EFEX_RTE_ORDR_ITEM_PRICE_MCU = 2205;
   public static final char EFEX_RTE_ORDR_ITEM_PRICE_RSU = 2206;
   public static final char EFEX_RTE_ORDR_ITEM_BRAND = 2207;
   public static final char EFEX_RTE_ORDR_ITEM_PACKSIZE = 2208;
   public static final char EFEX_RTE_ORDR_ITEM_REQUIRED = 2209;
   public static final char EFEX_RTE_ORDR_ITEM_UOM = 2210;
   public static final char EFEX_RTE_ORDR_ITEM_QTY = 2211;
   public static final char EFEX_RTE_ORDR_ITEM_VALUE = 2212;
   //
   public static final char EFEX_RTE_DISP_ITEM = 2301;
   public static final char EFEX_RTE_DISP_ITEM_ID = 2302;
   public static final char EFEX_RTE_DISP_ITEM_NAME = 2303;
   public static final char EFEX_RTE_DISP_ITEM_FLAG = 2304;
   //
   public static final char EFEX_RTE_ACTV_ITEM = 2401;
   public static final char EFEX_RTE_ACTV_ITEM_ID = 2402;
   public static final char EFEX_RTE_ACTV_ITEM_NAME = 2403;
   public static final char EFEX_RTE_ACTV_ITEM_FLAG = 2404;
   //
   public static final char EFEX_RTE_END = 2999;

   //
   // Customer constants
   //
   public static final char EFEX_CUS_STR = 3000;
   //
   public static final char EFEX_CUS = 3001;
   public static final char EFEX_CUS_DATA_TYPE = 3002;
   public static final char EFEX_CUS_DATA_ACTION = 3003;
   public static final char EFEX_CUS_CUSTOMER_ID = 3004;
   public static final char EFEX_CUS_CODE = 3016;
   public static final char EFEX_CUS_NAME = 3005;
   public static final char EFEX_CUS_STATUS = 3006;
   public static final char EFEX_CUS_ADDRESS = 3007;
   public static final char EFEX_CUS_CONTACT_NAME = 3008;
   public static final char EFEX_CUS_PHONE_NUMBER = 3009;
   public static final char EFEX_CUS_CUS_TYPE_ID = 3010;
   public static final char EFEX_CUS_OUTLET_LOCATION = 3011;
   public static final char EFEX_CUS_DISTRIBUTOR_ID = 3012;
   public static final char EFEX_CUS_POSTCODE = 3013;
   public static final char EFEX_CUS_FAX_NUMBER = 3014;
   public static final char EFEX_CUS_EMAIL_ADDRESS = 3015;
   //
   public static final char EFEX_CUS_END = 3999;

   //
   // Message constants
   //
   public static final char EFEX_MSG_STR = 4000;
   //
   public static final char EFEX_MSG = 4001;
   public static final char EFEX_MSG_ID = 4002;
   public static final char EFEX_MSG_OWNER = 4003;
   public static final char EFEX_MSG_TITLE = 4004;
   public static final char EFEX_MSG_TEXT = 4005;
   public static final char EFEX_MSG_STATUS = 4006;
   //
   public static final char EFEX_MSG_END = 4999;

   //
   // Reference product UOM constants
   //
   public static final char EFEX_UOM_STR = 5000;
   public static final char EFEX_UOM = 5001;
   public static final char EFEX_UOM_NAME = 5002;
   public static final char EFEX_UOM_TEXT = 5003;
   public static final char EFEX_UOM_END = 5999;

   //
   // Reference customer location constants
   //
   public static final char EFEX_CUS_LOCN_STR = 5100;
   public static final char EFEX_CUS_LOCN = 5101;
   public static final char EFEX_CUS_LOCN_NAME = 5102;
   public static final char EFEX_CUS_LOCN_TEXT = 5103;
   public static final char EFEX_CUS_LOCN_END = 5199;

   //
   // Reference customer type constants
   //
   public static final char EFEX_CUS_TYPE_STR = 5200;
   public static final char EFEX_CUS_TYPE = 5201;
   public static final char EFEX_CUS_TYPE_ID = 5202;
   public static final char EFEX_CUS_TYPE_NAME = 5203;
   public static final char EFEX_CUS_TYPE_CHANNEL_ID = 5204;
   public static final char EFEX_CUS_TYPE_END = 5299;

   //
   // Reference customer trade channel constants
   //
   public static final char EFEX_CUS_TRADE_CHANNEL_STR = 5300;
   public static final char EFEX_CUS_TRADE_CHANNEL = 5301;
   public static final char EFEX_CUS_TRADE_CHANNEL_ID = 5302;
   public static final char EFEX_CUS_TRADE_CHANNEL_NAME = 5303;
   public static final char EFEX_CUS_TRADE_CHANNEL_END = 5399;

   //
   // Reference distributor constants
   //
   public static final char EFEX_DIS_STR = 5400;
   public static final char EFEX_DIS = 5401;
   public static final char EFEX_DIS_ID = 5402;
   public static final char EFEX_DIS_NAME = 5403;
   public static final char EFEX_DIS_END = 5499;
   
   //
   // Class declarations
   //
   private Vector cobjMessages;

   /**
    * Constructs a new instance
    */
   public cMailbox() {
      cobjMessages = new Vector();
   }

   /**
    * Adds a new element to the collection
    * 
    * @param chrType the message type
    * @param strMessage the message data
    */
   public void addMessage(char chrType, String strData) {
      cMessage objMessage = new cMessage(chrType, strData);
      cobjMessages.addElement(objMessage);
   }
   
   /**
    * Adds bulk elements to the collection
    * 
    * @param objBuffer the mailbox string buffer
    * @exception Exception the fatal message
    */
   public void addBulkMessages(String objBuffer) throws Exception {
      char chrMessageCode = 0;
      short shrMessageLength = 0;
      String strMessageData = null;
      int i = 0;
      while (i < (objBuffer.length())) {
         chrMessageCode = objBuffer.charAt(i);
         i++;
         shrMessageLength = (short)objBuffer.charAt(i);
         i++;
         if (shrMessageLength <= 0) {
            strMessageData = "";
         } else {
            strMessageData = objBuffer.substring(i, (i + shrMessageLength));
            i = i + shrMessageLength;
         }
         this.addMessage(chrMessageCode, strMessageData);
      }
   }
   
   /**
    * Gets an existing element from the collection
    * 
    * @return cMessage the requested message reference
    * @param intIndex the index of the requested message
    * @exception Exception
    */
   public cMessage getMessage(int intIndex) throws Exception {
      return (cMessage)cobjMessages.elementAt(intIndex);
   }

   /**
    * Gets the count of the mailbox
    * 
    * @return int the mailbox message count
    */
   public int getMessageCount() {
      return cobjMessages.size();
   }
   
   /**
    * Puts the mailbox string buffer
    * 
    * @param objBuffer the mailbox string buffer
    * @exception Exception the fatal message
    */
   public void putBuffer(String objBuffer) throws Exception {
      if (objBuffer.length() < (BUFFER_OPEN.length() + BUFFER_CLOSE.length())) {
         throw new Exception("(Efex Mobile) Message buffer is incompatable with the application");
      }
      if (!objBuffer.substring(0, BUFFER_OPEN.length()).equals(BUFFER_OPEN)) {
         throw new Exception("(Efex Mobile) Message buffer open tag is incompatable with the application");
      }
      if (!objBuffer.substring(objBuffer.length() - BUFFER_CLOSE.length(), objBuffer.length()).equals(BUFFER_CLOSE)) {
         throw new Exception("(Efex Mobile) Message buffer close tag is incompatable with the application");
      }
      char chrMessageCode = 0;
      short shrMessageLength = 0;
      String strMessageData = null;
      int i = BUFFER_OPEN.length();
      while (i < (objBuffer.length() - BUFFER_CLOSE.length())) {
         chrMessageCode = objBuffer.charAt(i);
         i++;
         shrMessageLength = (short)objBuffer.charAt(i);
         i++;
         if (shrMessageLength <= 0) {
            strMessageData = "";
         } else {
            strMessageData = objBuffer.substring(i, (i + shrMessageLength));
            i = i + shrMessageLength;
         }
         this.addMessage(chrMessageCode, strMessageData);
      }
   }
   
   /**
    * Gets the mailbox string buffer
    * 
    * @return StringBuffer the mailbox string buffer
    * @exception Exception the fatal message
    */
   public StringBuffer getBuffer() throws Exception {
      cMessage objMessage;
      StringBuffer objBuffer = new StringBuffer(BUFFER_OPEN);
      for (int i=0;i<cobjMessages.size();i++) {
         objMessage = this.getMessage(i);
         objBuffer.append(objMessage.getMessageData());
      }
      objBuffer.append(BUFFER_CLOSE);
      return objBuffer;
   }
   
}

File: SOURCE/EFEX/BASE/MICROSOFT/C#/cMailbox.cs
/// <summary>
/// Type   : Class
/// Name   : cMailbox
/// Author : Softstep Pty Ltd
/// Date   : July 2008
/// </summary>
namespace EfexServer {
	
   using System;
   using System.Text;
   using System.Collections;
	
	/// <summary>
	/// This class defines a mailbox used for messaging
	/// </summary>
	public class cMailbox {
		
		//
      // Class constants
		//
      internal const string BUFFER_OPEN = "<EFEX>";
      internal const string BUFFER_CLOSE = "</EFEX>";

      //
      // Request constants
      //
      internal const char EFEX_RQS_USERNAME = (char)(101);
      internal const char EFEX_RQS_PASSWORD = (char)(102);
      internal const char EFEX_RQS_CUSTOMER_ID = (char)(103);
      internal const char EFEX_RQS_MESSAGE = (char)(199);

      //
      // Control constants
      //
      internal const char EFEX_CTL_STR = (char)(1000);
      //
      internal const char EFEX_CTL = (char)(1001);
      internal const char EFEX_CTL_USER_FIRSTNAME = (char)(1002);
      internal const char EFEX_CTL_USER_LASTNAME = (char)(1003);
      internal const char EFEX_CTL_MOBILE_DATE = (char)(1004);
      internal const char EFEX_CTL_MOBILE_STATUS = (char)(1005);
      internal const char EFEX_CTL_MOBILE_LOADED_TIME = (char)(1006);
      internal const char EFEX_CTL_MOBILE_SAVED_TIME = (char)(1007);
      //
      internal const char EFEX_CTL_END = (char)(1999);

      //
      // Route constants
      //
      internal const char EFEX_RTE_STR = (char)(2000);
      //
      internal const char EFEX_RTE_CALL = (char)(2001);
      internal const char EFEX_RTE_CALL_SEQUENCE = (char)(2002);
      internal const char EFEX_RTE_CALL_CUSTOMER_ID = (char)(2003);
      internal const char EFEX_RTE_CALL_CUSTOMER_CODE = (char)(2013);
      internal const char EFEX_RTE_CALL_CUSTOMER_NAME = (char)(2004);
      internal const char EFEX_RTE_CALL_CUSTOMER_TYPE = (char)(2005);
      internal const char EFEX_RTE_CALL_MARKET = (char)(2006);
      internal const char EFEX_RTE_CALL_STATUS = (char)(2007);
      internal const char EFEX_RTE_CALL_DATE = (char)(2008);
      internal const char EFEX_RTE_CALL_STR_TIME = (char)(2009);
      internal const char EFEX_RTE_CALL_END_TIME = (char)(2010);
      internal const char EFEX_RTE_CALL_ORDER_SEND = (char)(2011);
      internal const char EFEX_RTE_CALL_STOCK_DIST_COUNT = (char)(2012);
      //
      internal const char EFEX_RTE_STCK_ITEM = (char)(2101);
      internal const char EFEX_RTE_STCK_ITEM_ID = (char)(2102);
      internal const char EFEX_RTE_STCK_ITEM_NAME = (char)(2103);
      internal const char EFEX_RTE_STCK_ITEM_REQUIRED = (char)(2104);
      internal const char EFEX_RTE_STCK_ITEM_QTY = (char)(2105);
      //
      internal const char EFEX_RTE_ORDR_ITEM = (char)(2201);
      internal const char EFEX_RTE_ORDR_ITEM_ID = (char)(2202);
      internal const char EFEX_RTE_ORDR_ITEM_NAME = (char)(2203);
      internal const char EFEX_RTE_ORDR_ITEM_PRICE_TDU = (char)(2204);
      internal const char EFEX_RTE_ORDR_ITEM_PRICE_MCU = (char)(2205);
      internal const char EFEX_RTE_ORDR_ITEM_PRICE_RSU = (char)(2206);
      internal const char EFEX_RTE_ORDR_ITEM_BRAND = (char)(2207);
      internal const char EFEX_RTE_ORDR_ITEM_PACKSIZE = (char)(2208);
      internal const char EFEX_RTE_ORDR_ITEM_REQUIRED = (char)(2209);
      internal const char EFEX_RTE_ORDR_ITEM_UOM = (char)(2210);
      internal const char EFEX_RTE_ORDR_ITEM_QTY = (char)(2211);
      internal const char EFEX_RTE_ORDR_ITEM_VALUE = (char)(2212);
      //
      internal const char EFEX_RTE_DISP_ITEM = (char)(2301);
      internal const char EFEX_RTE_DISP_ITEM_ID = (char)(2302);
      internal const char EFEX_RTE_DISP_ITEM_NAME = (char)(2303);
      internal const char EFEX_RTE_DISP_ITEM_FLAG = (char)(2304);
      //
      internal const char EFEX_RTE_ACTV_ITEM = (char)(2401);
      internal const char EFEX_RTE_ACTV_ITEM_ID = (char)(2402);
      internal const char EFEX_RTE_ACTV_ITEM_NAME = (char)(2403);
      internal const char EFEX_RTE_ACTV_ITEM_FLAG = (char)(2404);
      //
      internal const char EFEX_RTE_END = (char)(2999);

      //
      // Customer constants
      //
      internal const char EFEX_CUS_STR = (char)(3000);
      //
      internal const char EFEX_CUS = (char)(3001);
      internal const char EFEX_CUS_DATA_TYPE = (char)(3002);
      internal const char EFEX_CUS_DATA_ACTION = (char)(3003);
      internal const char EFEX_CUS_CUSTOMER_ID = (char)(3004);
      internal const char EFEX_CUS_CODE = (char)(3016);
      internal const char EFEX_CUS_NAME = (char)(3005);
      internal const char EFEX_CUS_STATUS = (char)(3006);
      internal const char EFEX_CUS_ADDRESS = (char)(3007);
      internal const char EFEX_CUS_CONTACT_NAME = (char)(3008);
      internal const char EFEX_CUS_PHONE_NUMBER = (char)(3009);
      internal const char EFEX_CUS_CUS_TYPE_ID = (char)(3010);
      internal const char EFEX_CUS_OUTLET_LOCATION = (char)(3011);
      internal const char EFEX_CUS_DISTRIBUTOR_ID = (char)(3012);
      internal const char EFEX_CUS_POSTCODE = (char)(3013);
      internal const char EFEX_CUS_FAX_NUMBER = (char)(3014);
      internal const char EFEX_CUS_EMAIL_ADDRESS = (char)(3015);
      //
      internal const char EFEX_CUS_END = (char)(3999);

      //
      // Message constants
      //
      internal const char EFEX_MSG_STR = (char)(4000);
      //
      internal const char EFEX_MSG = (char)(4001);
      internal const char EFEX_MSG_ID = (char)(4002);
      internal const char EFEX_MSG_OWNER = (char)(4003);
      internal const char EFEX_MSG_TITLE = (char)(4004);
      internal const char EFEX_MSG_TEXT = (char)(4005);
      internal const char EFEX_MSG_STATUS = (char)(4006);
      //
      internal const char EFEX_MSG_END = (char)(4999);

      //
      // Reference product UOM constants
      //
      internal const char EFEX_UOM_STR = (char)(5000);
      internal const char EFEX_UOM = (char)(5001);
      internal const char EFEX_UOM_NAME = (char)(5002);
      internal const char EFEX_UOM_TEXT = (char)(5003);
      internal const char EFEX_UOM_END = (char)(5999);

      //
      // Reference customer location constants
      //
      internal const char EFEX_CUS_LOCN_STR = (char)(5100);
      internal const char EFEX_CUS_LOCN = (char)(5101);
      internal const char EFEX_CUS_LOCN_NAME = (char)(5102);
      internal const char EFEX_CUS_LOCN_TEXT = (char)(5103);
      internal const char EFEX_CUS_LOCN_END = (char)(5199);

      //
      // Reference customer type constants
      //
      internal const char EFEX_CUS_TYPE_STR = (char)(5200);
      internal const char EFEX_CUS_TYPE = (char)(5201);
      internal const char EFEX_CUS_TYPE_ID = (char)(5202);
      internal const char EFEX_CUS_TYPE_NAME = (char)(5203);
      internal const char EFEX_CUS_TYPE_CHANNEL_ID = (char)(5204);
      internal const char EFEX_CUS_TYPE_END = (char)(5299);

      //
      // Reference customer trade channel constants
      //
      internal const char EFEX_CUS_TRADE_CHANNEL_STR = (char)(5300);
      internal const char EFEX_CUS_TRADE_CHANNEL = (char)(5301);
      internal const char EFEX_CUS_TRADE_CHANNEL_ID = (char)(5302);
      internal const char EFEX_CUS_TRADE_CHANNEL_NAME = (char)(5303);
      internal const char EFEX_CUS_TRADE_CHANNEL_END = (char)(5399);

      //
      // Reference distributor constants
      //
      internal const char EFEX_DIS_STR = (char)(5400);
      internal const char EFEX_DIS = (char)(5401);
      internal const char EFEX_DIS_ID = (char)(5402);
      internal const char EFEX_DIS_NAME = (char)(5403);
      internal const char EFEX_DIS_END = (char)(5499);
		
		//
		// Class declarations
		//
		private ArrayList cobjMessages;
		
		/// <summary>
		/// Constructs a new instance
		/// </summary>
		internal cMailbox() {
			cobjMessages = new ArrayList();
		}
		
		/// <summary>
		/// Adds a new element to the collection
		/// </summary>
		/// <param name="chrCode">the message code</param>
		/// <param name="strData">the message data</param>
		internal void AddMessage(char chrCode, string strData) {
			cMessage objMessage = new cMessage(chrCode, strData);
			cobjMessages.Add(objMessage);
		}
		
		/// <summary>
		/// Gets an existing element from the collection
		/// </summary>
		/// <returns>cMessage the requested message reference</returns>
		/// <param name="intIndex">the index of the requested message</param>
		internal cMessage GetMessage(int intIndex) {
			return (cMessage)cobjMessages[intIndex];
		}
		
		/// <summary>
		/// Gets the count of the mailbox
		/// </summary>
		/// <returns>int the mailbox message count</returns>
      internal int GetMessageCount() {
         return cobjMessages.Count;
      }

		/// <summary>
		/// Puts the mailbox request string
		/// </summary>
		/// <param name="strRequest">the mailbox request string</param>
		internal void PutRequest(string strRequest) {
			if (strRequest.Length < (BUFFER_OPEN.Length + BUFFER_CLOSE.Length)) {
				throw new Exception("(Efex Server): Message buffer is incompatable with the application");
			}
			if (!strRequest.Substring(0, BUFFER_OPEN.Length).Equals(BUFFER_OPEN)) {
				throw new Exception("(Efex Server): Message buffer open tag is incompatable with the application");
			}
			if (!strRequest.Substring(strRequest.Length - BUFFER_CLOSE.Length, BUFFER_CLOSE.Length).Equals(BUFFER_CLOSE)) {
				throw new Exception("(Efex Server): Message buffer close tag is incompatable with the application");
			}
			char chrMessageCode = (char)(0);
			short shrMessageLength = 0;
         string strMessageData = null;
			int i = BUFFER_OPEN.Length;
			while (i < (strRequest.Length - BUFFER_CLOSE.Length)) {
				chrMessageCode = (char)strRequest[i];
				i++;
				shrMessageLength = (short)strRequest[i];
				i++;
				if (shrMessageLength <= 0) {
               strMessageData = null;
				} else {
               strMessageData = strRequest.Substring(i, (int)shrMessageLength);
					i = i + shrMessageLength;
				}
            this.AddMessage(chrMessageCode, strMessageData);
			}
		}
		
		/// <summary>
		/// Gets the mailbox response string
		/// </summary>
		/// <returns>string the mailbox response string</returns>
		internal string GetResponse() {  
         StringBuilder objBuffer = new StringBuilder(BUFFER_OPEN);
         cMessage objMessage;
         for (int i=0; i<cobjMessages.Count; i++) {
            objMessage = (cMessage)cobjMessages[i];
            objBuffer.Append(objMessage.GetMessageData());
         }
         objMessage = null;
         objBuffer.Append(BUFFER_CLOSE);
         return objBuffer.ToString();
      }

	}

}